src/rust_lexer/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use std::str::Chars;
+
+pub struct Token {
+    pub kind: TokenKind,
+    pub len: usize,
+}
+
+
+pub enum TokenKind {
+    Shebang,
+    LineComment,
+    BlockComment {
+        unterminated: bool,
+    },
+    TODO,
+}
+use self::TokenKind::*;
+
+impl Token {
+    fn new(kind: TokenKind, len: usize) -> Token {
+        Token { kind, len }
+    }
+}
+
+pub fn next_token(input: &str, is_beginning_of_file: bool) -> Option<Token> {
```

Comment:
I would call this `match_start` or similar (could be a method on `Token`, even?)

Now that I think about it, `is_beginning_of_file` is only used for shebangs, right?

I don't think those should be processed here, but rather there should be a separate method to strip shebangs (and therefore not have to check them all the time).

That would make this a `&str -> Option<Self>`, which is very close to `FromStr`!
I think that would make sense, `proc_macro` already uses `FromStr` (for lexing into a `TokenStream`).